[gloo] dont hold locks in calls to buffer in ProcessGroupGloo:RecvWork::wait() and
SendWork::wait()

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33926 [gloo] dont hold locks in calls to buffer in ProcessGroupGloo:RecvWork::wait() and
SendWork::wait()**
SendWork::wait()**

Closes https://github.com/pytorch/pytorch/issues/30164
The UnboundBuffer calls here are already protected by a mutex. We only
need to hold the lock while writing the shared structures completed_ and
exception_.

Differential Revision: [D20154546](https://our.internmc.facebook.com/intern/diff/D20154546/)

rohan-varma committed:
[gloo] dont hold locks in calls to buffer in ProcessGroupGloo:RecvWork::wait() and
SendWork::wait()

The UnboundBuffer calls here are already protected by a mutex. We only
need to hold the lock while writing the shared structures completed_ and
exception_.

Differential Revision: [D20154546](https://our.internmc.facebook.com/intern/diff/D20154546/)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
Let's make them lock guard to be explicit that the following code will not manually unlock them?

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
Can we avoid having two locks here by having a function local exception variable and only acquire the lock after the catch? Will it be better if we have a finally block and acquire lock there? sth like:

```c++
try {
  sendCompleted = buffer_->waitSend();
} catch (...) {
  exception = std::current_exception();
} finally {
  std::lock_guard<std::mutex> lock(mutex_);
  completed_ = true;
  if (exception) {
    std::swap(exception, exception_);
    std::rethrow_exception(exception_);
  }
}

return sendCompleted;
```

mrshenli reviewed (approved):
Thanks for fixing! Left some minor comments inline. Stamp to unblock.

rohan-varma commented on torch/lib/c10d/ProcessGroupGloo.cpp:
I think we can add the function local variable, although C++ does not support `finally` for the 2nd suggestion.

rohan-varma committed:
Update on "[gloo] dont hold locks in calls to buffer in ProcessGroupGloo:RecvWork::wait() and
SendWork::wait()"


Closes https://github.com/pytorch/pytorch/issues/30164
The UnboundBuffer calls here are already protected by a mutex. We only
need to hold the lock while writing the shared structures completed_ and
exception_.

Differential Revision: [D20154546](https://our.internmc.facebook.com/intern/diff/D20154546/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[gloo] dont hold locks in calls to buffer in ProcessGroupGloo:RecvWork::wait() and
SendWork::wait()"

SendWork::wait()**

Closes https://github.com/pytorch/pytorch/issues/30164
The UnboundBuffer calls here are already protected by a mutex. We only
need to hold the lock while writing the shared structures completed_ and
exception_.

Differential Revision: [D20154546](https://our.internmc.facebook.com/intern/diff/D20154546/)

[ghstack-poisoned]
